games/maze/src/maze.rs

```diff
@@ -57,24 +66,120 @@ pub enum Direction {
 
 impl Maze {
     pub fn new(size: usize) -> Self {
+        let random_map = Maze::generate_random_map(size);
+
         let mut maze = Maze {
             player: Position { x: 0, y: 0 },
             exit: Position {
                 x: size - 1,
                 y: size - 1,
             },
             size,
-            map: vec![Tile::blocked(); size * size],
+            map: random_map,
         };
 
+        maze.reveal_around_player();
+        maze
+    }
+
+    fn generate_random_map(size: usize) -> Vec<Tile> {
+        fn find(
+            size: usize,
+            map: &[MazeGenerationTile],
+            (p, q): (Position, Position),
+        ) -> (Position, Position) {
+            let (cell_p, cell_q) = (map[size * p.y + p.x].link, map[size * q.y + q.x].link);
+
+            if p != cell_p || q != cell_q {
+                find(size, map, (cell_p, cell_q))
+            } else {
+                (cell_p, cell_q)
+            }
+        }
+
+        if size & 1 == 0 {
+            panic!("Random maze only allows odd numbers")
+        };
+
+        let mut gen_map = Vec::with_capacity(size * size);
+
         for i in 0..size {
-            maze.set(i, 0, Tile::open());
-            maze.set(size - 1, i, Tile::open());
+            for j in 0..size {
+                let pos = Position { x: j, y: i };
+
+                match (j & 1 == 0, i & 1 == 0) {
+                    (true, true) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: Some(TileType::Open),
+                    }),
+                    (false, false) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: Some(TileType::Blocked),
+                    }),
+                    (false, true) | (true, false) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: None,
+                    }),
+                }
+            }
         }
 
-        maze.reveal_around_player();
+        let neither_map = gen_map.to_vec();
+        let mut neither_map = neither_map
+            .iter()
```

Comment:
`iter()` only borrows the Vec it is called on and the collect below will be creating a new Vec as well so I do not think you need to clone `gen_map` with `to_vec()` here. `gen_map.iter()....collect()` should be enough and reduce one allocation and mem copy.